--- a/net/wireless/util.c	2021-06-06 22:16:07.321840000 +0800
+++ b/net/wireless/util.c	2026-09-05 20:45:57.761640832 +0800
@@ -76,13 +76,13 @@
 {
 	/* see 802.11 17.3.8.3.2 and Annex J
 	 * there are overlapping channel numbers in 5GHz and 2GHz bands */
-	if (chan <= 0)
+	if (chan <= 0 && !(band == NL80211_BAND_2GHZ && chan >= -19))
 		return 0; /* not supported */
 	switch (band) {
 	case NL80211_BAND_2GHZ:
 		if (chan == 14)
 			return MHZ_TO_KHZ(2484);
-		else if (chan < 14)
+		else if (chan != 0 && chan <= 65)
 			return MHZ_TO_KHZ(2407 + chan * 5);
 		break;
 	case NL80211_BAND_5GHZ:
@@ -148,6 +148,9 @@
 		return 14;
 	else if (freq < 2484)
 		return (freq - 2407) / 5;
+	else if (freq <= 2732)
+		/* extended 2.4 GHz band: 2487..2732 MHz -> ch 16..65 */
+		return (freq - 2407) / 5;
 	else if (freq >= 4910 && freq <= 4980)
 		return (freq - 4000) / 5;
 	else if (freq < 5925)
