--- a/feeds/luci/modules/luci-mod-admin-full/luasrc/model/cbi/admin_network/wifi.lua
+++ b/feeds/luci/modules/luci-mod-admin-full/luasrc/model/cbi/admin_network/wifi.lua
@@ -108,6 +108,7 @@
 s:tab("general", translate("General Setup"))
 s:tab("macfilter", translate("MAC-Filter"))
 s:tab("advanced", translate("Advanced Settings"))
+s:tab("rssi", translate("low RSSI disconnect"))
 
 --[[
 back = s:option(DummyValue, "_overview", translate("Overview"))
@@ -360,6 +361,7 @@
 s:tab("encryption", translate("Wireless Security"))
 s:tab("macfilter", translate("MAC-Filter"))
 s:tab("advanced", translate("Advanced Settings"))
+s:tab("rssi", translate("low RSSI disconnect"))
 
 s:taboption("general", Value, "ssid", translate("<abbr title=\"Extended Service Set Identifier\">ESSID</abbr>"))
 
@@ -473,6 +475,50 @@
 	wmm:depends({mode="ap"})
 	wmm:depends({mode="ap-wds"})
 	wmm.default = wmm.enabled
+
+	signal_connect = s:taboption("rssi", Value, "signal_connect",
+		translate("signal_connect"),
+		translate("signal accessible value"))
+	signal_connect:depends({mode="ap"},{mode="ap-wds"},{mode="mesh"})
+	signal_connect.placeholder = "-95"
+	--signal_connect.datatype = "range(-10,-95)"
+	signal_connect.default = "-95"
+
+	signal_stay = s:taboption("rssi", Value, "signal_stay",
+		translate("signal_stay"),
+		translate("signal automatic disconnection value"))
+	signal_stay:depends({mode="ap"},{mode="ap-wds"},{mode="mesh"})
+	signal_stay.placeholder = "-100"
+	--signal_stay.datatype = "range(-30,-100)"
+	signal_stay.default = "-100"
+
+	signal_strikes = s:taboption("rssi", Value, "signal_strikes",
+		translate("signal_strikes"),
+		translate("signal strikes times"))
+	signal_strikes:depends({mode="ap"},{mode="ap-wds"},{mode="mesh"})
+	signal_strikes.placeholder = "3"
+	signal_strikes.datatype = "range(1,10)"
+	signal_strikes.default = "3"
+
+	signal_poll_time = s:taboption("rssi", Value, "signal_poll_time",
+		translate("signal_poll_time"),
+		translate("signal polling times"))
+	signal_poll_time:depends({mode="ap"},{mode="ap-wds"},{mode="mesh"})
+	signal_poll_time.placeholder = "5"
+	signal_poll_time.datatype = "range(3,20)"
+	signal_poll_time.default = "5"
+
+	signal_drop_reason = s:taboption("rssi", Value, "signal_drop_reason",
+		translate("signal_drop_reason"),
+		translate("signal drop reason"))
+	signal_drop_reason:depends({mode="ap"},{mode="ap-wds"},{mode="mesh"})
+	signal_drop_reason.placeholder = "3"
+	signal_drop_reason.datatype = "range(3,5)"
+	signal_drop_reason.default = "3"
+
+	isolate = s:taboption("advanced", Flag, "isolate", translate("Separate Clients"),
+	 translate("Prevents client-to-client communication"))
+	isolate:depends({mode="ap"})
 end
 
 
--- a/feeds/luci/modules/luci-base/po/zh-cn/base.po
+++ b/feeds/luci/modules/luci-base/po/zh-cn/base.po
@@ -3157,6 +3157,25 @@
 msgid "« Back"
 msgstr "« 后退"
 
+
+msgid "low RSSI disconnect"
+msgstr "弱信号剔除"
+
+msgid "signal accessible value"
+msgstr "允许接入的信号值（当单台设备使用时，建议设定为-95dBm。这样可以防止客户端被BAN。当多设备做无线覆盖时，请一定把每台机器的SSID设定成相同名称，并推荐使用-70dBm。默认-95）"
+
+msgid "signal automatic disconnection value"
+msgstr "自动剔除弱信号（当单台设备使用时，建议设定为-100dBm。当多设备做无线覆盖时，推荐使用-75dBm。默认-100）"
+
+msgid "signal strikes times"
+msgstr "信号检测次数（默认3，建议使用默认值）"
+
+msgid "signal polling times"
+msgstr "信号轮询次数（默认5，建议使用默认值）"
+
+msgid "signal drop reason"
+msgstr "信号检测次数（默认3，建议使用默认值）"
+
 #~ msgid "CPU frequency"
 #~ msgstr "CPU 频率"
 
