--- a/feeds/packages/libs/libssh2/Makefile
+++ b/feeds/packages/libs/libssh2/Makefile
@@ -1,5 +1,5 @@
 #
-# Copyright (C) 2015 - 2018 OpenWrt.org
+# Copyright (C) 2015-2018 OpenWrt.org
 #
 # This is free software, licensed under the GNU General Public License v2.
 # See /LICENSE for more information.
@@ -8,52 +8,60 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=libssh2
-PKG_VERSION:=1.6.0
+PKG_VERSION:=1.9.0
 PKG_RELEASE:=2
 
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
 PKG_SOURCE_URL:=https://www.libssh2.org/download
-PKG_MD5SUM:=00aabd6e714a5f42a4fb82ace20db1dd
+PKG_MD5SUM:=1beefafe8963982adc84b408b2959927
 
-PKG_INSTALL:=1
-
-PKG_LICENSE:=BSD
+PKG_MAINTAINER:=Jiri Slachta <jiri@slachta.eu>
+PKG_LICENSE:=BSD-3-Clause
 PKG_LICENSE_FILES:=COPYING
+PKG_CPE_ID:=cpe:/a:libssh2:libssh2
+
+CMAKE_INSTALL:=1
+PKG_BUILD_PARALLEL:=1
+
+PKG_CONFIG_DEPENDS:= \
+	CONFIG_LIBSSH2_MBEDTLS \
+	CONFIG_LIBSSH2_OPENSSL
 
 include $(INCLUDE_DIR)/package.mk
+include $(INCLUDE_DIR)/cmake.mk
 
 define Package/libssh2
   SECTION:=libs
   CATEGORY:=Libraries
   TITLE:=SSH2 library
   URL:=https://www.libssh2.org/
-  DEPENDS:=+libopenssl +zlib
-  MAINTAINER:=Jiri Slachta <jiri@slachta.eu>
+  DEPENDS:=+LIBSSH2_MBEDTLS:libmbedtls +LIBSSH2_OPENSSL:libopenssl +zlib
+  ABI_VERSION:=1
 endef
 
 define Package/libssh2/description
  libssh2 is a client-side C library implementing the SSH2 protocol.
 endef
 
-TARGET_CFLAGS += $(FPIC)
-
-CONFIGURE_ARGS += \
-	--disable-examples-build \
-	--with-libssl-prefix=$(STAGING_DIR)/usr \
-	--with-libz-prefix=$(STAGING_DIR)/usr
-
-define Build/InstallDev
-	$(INSTALL_DIR) $(1)/usr/include
-	$(INSTALL_DIR) $(1)/usr/lib
-	$(INSTALL_DIR) $(1)/usr/lib/pkgconfig
-	$(CP) $(PKG_INSTALL_DIR)/usr/include/*.h $(1)/usr/include/
-	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libssh2.so* $(1)/usr/lib/
-	$(CP) $(PKG_INSTALL_DIR)/usr/lib/pkgconfig/libssh2.pc $(1)/usr/lib/pkgconfig/
+define Package/libssh2/config
+	source "$(SOURCE)/Config.in"
 endef
 
+CMAKE_OPTIONS += \
+	-DBUILD_SHARED_LIBS=ON \
+	-DBUILD_TESTING=OFF \
+	-DENABLE_ZLIB_COMPRESSION=ON \
+	-DCLEAR_MEMORY=ON
+
+ifeq ($(CONFIG_LIBSSH2_OPENSSL),y)
+	CMAKE_OPTIONS += -DCRYPTO_BACKEND=OpenSSL
+else
+	CMAKE_OPTIONS += -DCRYPTO_BACKEND=mbedTLS
+endif
+
 define Package/libssh2/install
 	$(INSTALL_DIR) $(1)/usr/lib
-	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libssh2.so* $(1)/usr/lib/
+	$(CP) $(PKG_INSTALL_DIR)/usr/lib/libssh2.so.$(ABI_VERSION)* $(1)/usr/lib/
 endef
 
 $(eval $(call BuildPackage,libssh2))
--- a/feeds/packages/libs/libssh2/patches/0004-CVE-2016-0787.patch
+++ /dev/null
@@ -1,22 +0,0 @@
-Description: CVE-2016-0787: Truncated Difffie-Hellman secret length
- Convert bytes to bits in diffie_hellman_sha1. Otherwise we get far too
- small numbers.
-Origin: backport, http://www.libssh2.org/CVE-2016-0787.patch
-Forwarded: not-needed
-Author: Daniel Stenberg <daniel@haxx.se>
-Reviewed-by: Salvatore Bonaccorso <carnil@debian.org>
-Last-Update: 2016-02-18
-Applied-Upstream: 1.7.0
----
-
---- a/src/kex.c
-+++ b/src/kex.c
-@@ -103,7 +103,7 @@ static int diffie_hellman_sha1(LIBSSH2_S
-         memset(&exchange_state->req_state, 0, sizeof(packet_require_state_t));
- 
-         /* Generate x and e */
--        _libssh2_bn_rand(exchange_state->x, group_order, 0, -1);
-+        _libssh2_bn_rand(exchange_state->x, group_order * 8 - 1, 0, -1);
-         _libssh2_bn_mod_exp(exchange_state->e, g, exchange_state->x, p,
-                             exchange_state->ctx);
- 
--- /dev/null
+++ a/feeds/packages/libs/libssh2/Config.in
@@ -0,0 +1,15 @@
+if PACKAGE_libssh2
+
+choice
+	prompt "Choose crypto backend"
+	default LIBSSH2_OPENSSL
+
+	config LIBSSH2_MBEDTLS
+	bool "mbedtls"
+
+	config LIBSSH2_OPENSSL
+	bool "openssl"
+
+endchoice
+
+endif
