--- a/src/mrib.c
+++ b/src/mrib.c
@@ -383,8 +383,6 @@ static void mrib_receive_mrt6(struct ulo
 			inet_ntop(AF_INET6, &from.sin6_addr, addrbuf, sizeof(addrbuf));
 
 			if (!IN6_IS_ADDR_LINKLOCAL(&from.sin6_addr) || hlim != 1 || len < 24 || !alert) {
-				L_WARN("mld: ignoring invalid MLD-message of type %d from %s on %d",
-						mld->mld_icmp6_hdr.icmp6_type, addrbuf, ifindex);
 				continue;
 			}
 
--- a/src/groups.c
+++ b/src/groups.c
@@ -260,7 +260,6 @@ void groups_update_timers(struct groups
 	inet_ntop(AF_INET6, groupaddr, addrbuf, sizeof(addrbuf));
 	struct group *group = groups_get_group(groups, groupaddr, NULL);
 	if (!group) {
-		L_WARN("%s: failed to update timer: no such group %s", __FUNCTION__, addrbuf);
 		return;
 	}
 
@@ -275,7 +274,6 @@ void groups_update_timers(struct groups
 		for (size_t i = 0; i < len; ++i) {
 			struct group_source *source = groups_get_source(groups, group, &addrs[i], NULL);
 			if (!source) {
-				L_WARN("%s: failed to update timer: unknown sources for group %s", __FUNCTION__, addrbuf);
 				continue;
 			}
 
