--- a/backport-include/linux/pci.h
+++ b/backport-include/linux/pci.h
@@ -7,10 +7,6 @@
 #include <linux/pci-aspm.h>
 #endif
 
-#if LINUX_VERSION_IS_LESS(5,4,0)
-#include <linux/pci-aspm.h>
-#endif
-
 #ifndef module_pci_driver
 /**
  * module_pci_driver() - Helper macro for registering a PCI driver
@@ -265,29 +261,4 @@
 #endif /* < 5.3 */
 #endif /* defined(CONFIG_PCI) */
 
-#if defined(CONFIG_PCI) && LINUX_VERSION_IS_LESS(5,3,0)
-
-static inline int
-LINUX_BACKPORT(pci_disable_link_state)(struct pci_dev *pdev, int state)
-{
-	u16 aspmc;
-
-	pci_disable_link_state(pdev, state);
-
-	pcie_capability_read_word(pdev, PCI_EXP_LNKCTL, &aspmc);
-	if ((state & PCIE_LINK_STATE_L0S) &&
-	    (aspmc & PCI_EXP_LNKCTL_ASPM_L0S))
-		return -EPERM;
-
-	if ((state & PCIE_LINK_STATE_L1) &&
-	    (aspmc & PCI_EXP_LNKCTL_ASPM_L1))
-		return -EPERM;
-
-	return 0;
-}
-
-#define pci_disable_link_state LINUX_BACKPORT(pci_disable_link_state)
-
-#endif
-
 #endif /* _BACKPORT_LINUX_PCI_H */
