--- a/station.c
+++ b/station.c
@@ -7,7 +7,6 @@
 #include <netlink/genl/ctrl.h>
 #include <netlink/msg.h>
 #include <netlink/attr.h>
-#include <time.h>
 
 #include "nl80211.h"
 #include "iw.h"
@@ -594,24 +593,8 @@ static int print_sta_handler(struct nl_m
 	if (sinfo[NL80211_STA_INFO_CONNECTED_TIME])
 		printf("\n\tconnected time:\t%u seconds",
 			nla_get_u32(sinfo[NL80211_STA_INFO_CONNECTED_TIME]));
-	if (sinfo[NL80211_STA_INFO_ASSOC_AT_BOOTTIME]) {
-		unsigned long long bt;
-		struct timespec now_ts;
-		unsigned long long boot_ns;
-		unsigned long long assoc_at_ms;
 
-		clock_gettime(CLOCK_BOOTTIME, &now_ts);
-		boot_ns = now_ts.tv_sec * 1000000000ULL;
-		boot_ns += now_ts.tv_nsec;
-
-		bt = (unsigned long long)nla_get_u64(sinfo[NL80211_STA_INFO_ASSOC_AT_BOOTTIME]);
-		printf("\n\tassociated at [boottime]:\t%llu.%.3llus",
-		       bt/1000000000, (bt%1000000000)/1000000);
-		assoc_at_ms = now_ms - ((boot_ns - bt) / 1000000);
-		printf("\n\tassociated at:\t%llu ms", assoc_at_ms);
-	}
-
-	printf("\n\tcurrent time:\t%llu ms\n", now_ms);
+	printf("\n");
 	return NL_SKIP;
 }
 
