--- a/vars.example
+++ b/vars.example
@@ -101,7 +101,7 @@
 # generation take much longer. Values up to 4096 should be accepted by most
 # software. Only used when the crypto alg is rsa (see below.)
 
-#set_var EASYRSA_KEY_SIZE	2048
+set_var EASYRSA_KEY_SIZE	1024
 
 # The default crypto mode is rsa; ec can enable elliptic curve support.
 # Note that not all software supports ECC, so use care when enabling it.
