--- a/mt7615/Kconfig
+++ b/mt7615/Kconfig
@@ -38,3 +38,16 @@
 	  This adds support for MT7663U 802.11ax 2x2:2 wireless devices.
 
 	  To compile this driver as a module, choose M here.
+
+config MT7615_OFFLOAD_FIRMWARE
+	bool "Prefer client mode offload firmware (MT7663)"
+	depends on MT7615E || MT7663U
+	default y
+	help
+	  Load MT7663 client mode offload firmware (v3) as primary option
+	  and fallback to MT7663 firmware v2 in case of failure.
+	  If MT7615_OFFLOAD_FIRMWARE is not selected MT7663 firmware v2
+	  will be used as primary option.
+	  MT7663 client mode offload firmware supports low power features
+	  (hw frequency scanning, scheduled frequency scanning, WoW,
+	  802.11 power save) but is more limited in AP mode
--- a/mt7615/mcu.c
+++ b/mt7615/mcu.c
@@ -12,11 +12,6 @@
 #include "mac.h"
 #include "eeprom.h"
 
-static bool prefer_offload_fw = true;
-module_param(prefer_offload_fw, bool, 0644);
-MODULE_PARM_DESC(prefer_offload_fw,
-		 "Prefer client mode offload firmware (MT7663)");
-
 struct mt7615_patch_hdr {
 	char build_date[16];
 	char platform[4];
@@ -2152,7 +2147,7 @@
 	const char *primary_rom = MT7663_OFFLOAD_ROM_PATCH;
 	int ret;
 
-	if (!prefer_offload_fw) {
+	if (!IS_ENABLED(CONFIG_MT7615_OFFLOAD_FIRMWARE)) {
 		secondary_rom = MT7663_OFFLOAD_ROM_PATCH;
 		primary_rom = MT7663_ROM_PATCH;
 	}
