--- a/mt7603/mac.c
+++ b/mt7603/mac.c
@@ -1442,6 +1442,8 @@
 		mt76_queue_rx_reset(dev, i);
 	}
 
+	mt76_tx_status_check(&dev->mt76, NULL, true);
+
 	mt7603_dma_sched_reset(dev);
 
 	mt7603_mac_dma_start(dev);
--- a/mt76x02_mmio.c
+++ b/mt76x02_mmio.c
@@ -490,6 +490,8 @@
 		mt76_queue_rx_reset(dev, i);
 	}
 
+	mt76_tx_status_check(&dev->mt76, NULL, true);
+
 	mt76x02_mac_start(dev);
 
 	if (dev->ed_monitor)
