--- a/dma.c
+++ b/dma.c
@@ -604,6 +604,7 @@
 	dev = container_of(napi->dev, struct mt76_dev, napi_dev);
 	qid = napi - dev->napi;

+	local_bh_disable();
 	rcu_read_lock();

 	do {
@@ -613,6 +614,7 @@
 	} while (cur && done < budget);

 	rcu_read_unlock();
+	local_bh_enable();

 	if (done < budget) {
 		napi_complete(napi);
@@ -630,8 +632,8 @@
 	init_dummy_netdev(&dev->napi_dev);

 	mt76_for_each_q_rx(dev, i) {
-		netif_threaded_napi_add(&dev->napi_dev, &dev->napi[i],
-					mt76_dma_rx_poll, 64);
+		netif_napi_add(&dev->napi_dev, &dev->napi[i], mt76_dma_rx_poll,
+			       64);
 		mt76_dma_rx_fill(dev, &dev->q_rx[i]);
 		napi_enable(&dev->napi[i]);
 	}
