Index: compat-wireless-2016-01-10/net/mac80211/tx.c
===================================================================
--- compat-wireless-2016-01-10.orig/net/mac80211/tx.c	2017-03-28 17:27:03.333271558 +0800
+++ compat-wireless-2016-01-10/net/mac80211/tx.c	2017-03-28 17:44:08.289301776 +0800
@@ -627,6 +627,9 @@
 		case WLAN_CIPHER_SUITE_CCMP_256:
 		case WLAN_CIPHER_SUITE_GCMP:
 		case WLAN_CIPHER_SUITE_GCMP_256:
+#ifdef CPTCFG_WLAN_VENDOR_RWNX
+        case WLAN_CIPHER_SUITE_SMS4:
+#endif
 			if (!ieee80211_is_data_present(hdr->frame_control) &&
 			    !ieee80211_use_mfp(hdr->frame_control, tx->sta,
 					       tx->skb))
