--- a/net/mac80211/rc80211_minstrel_ht.c
+++ b/net/mac80211/rc80211_minstrel_ht.c
@@ -1210,6 +1210,15 @@
 			ldpc = true;
 	}
 
+	if (ht_cap & IEEE80211_HT_CAP_LDPC_CODING) {
+		if (use_vht) {
+			if (vht_cap->cap & IEEE80211_VHT_CAP_RXLDPC)
+				mi->tx_flags |= IEEE80211_TX_CTL_LDPC;
+		} else {
+			mi->tx_flags |= IEEE80211_TX_CTL_LDPC;
+		}
+	}
+
 	mi->tx_flags |= stbc << IEEE80211_TX_CTL_STBC_SHIFT;
 	if (ldpc)
 		mi->tx_flags |= IEEE80211_TX_CTL_LDPC;
