Index: compat-wireless-2016-01-10/net/mac80211/rx.c
===================================================================
--- compat-wireless-2016-01-10.orig/net/mac80211/rx.c	2017-10-31 15:10:55.550640036 +0800
+++ compat-wireless-2016-01-10/net/mac80211/rx.c	2017-10-31 15:11:37.906641284 +0800
@@ -2232,7 +2232,7 @@
 	struct ieee80211_local *local = rx->local;
 	struct ieee80211_sub_if_data *sdata = rx->sdata;
 	struct ieee80211_if_mesh *ifmsh = &sdata->u.mesh;
-	u16 ac, q, hdrlen;
+	u16 q, hdrlen;
 
 	hdr = (struct ieee80211_hdr *) skb->data;
 	hdrlen = ieee80211_hdrlen(hdr->frame_control);
@@ -2302,9 +2302,8 @@
 	    ether_addr_equal(sdata->vif.addr, hdr->addr3))
 		return RX_CONTINUE;
 
-	ac = ieee80211_select_queue_80211(sdata, skb, hdr);
-	q = sdata->vif.hw_queue[ac];
-	if (ieee80211_queue_stopped(&local->hw, q)) {
+	q = ieee80211_select_queue_80211(sdata, skb, hdr);
+	if (ieee80211_queue_stopped(&local->hw, sdata->vif.hw_queue[q])) {
 		IEEE80211_IFSTA_MESH_CTR_INC(ifmsh, dropped_frames_congestion);
 		return RX_DROP_MONITOR;
 	}
