Index: compat-wireless-2016-01-10/include/net/cfg80211.h
===================================================================
--- compat-wireless-2016-01-10.orig/include/net/cfg80211.h	2019-04-16 16:58:22.309544151 +0800
+++ compat-wireless-2016-01-10/include/net/cfg80211.h	2019-04-16 17:00:27.333547837 +0800
@@ -487,6 +487,13 @@
 			     u32 prohibited_flags);
 
 /**
+ * cfg80211_chandef_dfs_available - checks if radar channel is available
+ * returns: true available, false not available.
+ */
+bool cfg80211_chandef_dfs_available(struct wiphy *wiphy,
+			const struct cfg80211_chan_def *chandef);
+
+/**
  * cfg80211_chandef_dfs_required - checks if radar detection is required
  * @wiphy: the wiphy to validate against
  * @chandef: the channel definition to check
Index: compat-wireless-2016-01-10/net/mac80211/chan.c
===================================================================
--- compat-wireless-2016-01-10.orig/net/mac80211/chan.c	2016-01-15 20:00:10.000000000 +0800
+++ compat-wireless-2016-01-10/net/mac80211/chan.c	2019-04-16 17:00:16.177547508 +0800
@@ -1601,7 +1601,10 @@
 	ieee80211_recalc_smps_chanctx(local, ctx);
 	ieee80211_recalc_radar_chanctx(local, ctx);
  out:
-	if (ret)
+	/* we call cfg80211_chandef_dfs_available here to set radar_required false
+	 * when chandef is available, or we can't scan on radar channel.
+	 */
+	if (ret || cfg80211_chandef_dfs_available(local->hw.wiphy, chandef))
 		sdata->radar_required = false;
 
 	mutex_unlock(&local->chanctx_mtx);
Index: compat-wireless-2016-01-10/net/wireless/chan.c
===================================================================
--- compat-wireless-2016-01-10.orig/net/wireless/chan.c	2019-04-16 16:58:23.009544172 +0800
+++ compat-wireless-2016-01-10/net/wireless/chan.c	2019-04-16 17:00:01.981547090 +0800
@@ -487,7 +487,7 @@
 	return true;
 }
 
-static bool cfg80211_chandef_dfs_available(struct wiphy *wiphy,
+bool cfg80211_chandef_dfs_available(struct wiphy *wiphy,
 				const struct cfg80211_chan_def *chandef)
 {
 	int width;
@@ -520,6 +520,7 @@
 
 	return r;
 }
+EXPORT_SYMBOL(cfg80211_chandef_dfs_available);
 
 static unsigned int cfg80211_get_chans_dfs_cac_time(struct wiphy *wiphy,
 						    u32 center_freq,
