--- a/src/https_client.c
+++ b/src/https_client.c
@@ -169,7 +169,7 @@
       }
     }
 
-    ELOG("%4.4lx: %s%s", (long)i, hex, str);
+    /* ELOG("%4.4lx: %s%s", (long)i, hex, str); */
   }
 }
 
@@ -188,9 +188,9 @@
     if (long_resp == 200) {
       faulty_response = 0;
     } else if (long_resp == 0) {
-      ELOG("No response");
+      /* ELOG("No response"); */
     } else {
-      ELOG("curl response code: %d, content length: %zu", long_resp, ctx->buflen);
+      /* ELOG("curl response code: %d, content length: %zu", long_resp, ctx->buflen); */
       if (ctx->buflen >= 0) {
         https_log_response_content(ctx->buf, ctx->buflen);
       }
@@ -225,7 +225,7 @@
             ctx->curl, CURLINFO_SSL_VERIFYRESULT, &long_resp)) != CURLE_OK) {
       ELOG("CURLINFO_SSL_VERIFYRESULT: %s", curl_easy_strerror(res));
     } else if (long_resp != CURLE_OK) {
-      ELOG("CURLINFO_SSL_VERIFYRESULT: %s", curl_easy_strerror(long_resp));
+      /* ELOG("CURLINFO_SSL_VERIFYRESULT: %s", curl_easy_strerror(long_resp)); */
     }
     if ((res = curl_easy_getinfo(
             ctx->curl, CURLINFO_OS_ERRNO, &long_resp)) != CURLE_OK) {
